pkg/controller/workload/job/job_controller.go

```diff
@@ -432,29 +465,41 @@ func ConstructWorkloadFor(ctx context.Context, client client.Client,
 	return w, nil
 }
 
-func appendFinishedConditionIfNotExists(conds []metav1.Condition, jobStatus batchv1.JobConditionType) ([]metav1.Condition, bool) {
-	for i, c := range conds {
-		if c.Type == kueue.WorkloadFinished {
-			if c.Status == metav1.ConditionTrue {
-				return conds, false
-			}
-			conds = append(conds[:i], conds[i+1:]...)
-			break
-		}
+func podsCount(jobSpec *batchv1.JobSpec) int32 {
+	// parallelism is always set as it is otherwise defaulted by k8s to 1
+	podsCount := *(jobSpec.Parallelism)
+	if jobSpec.Completions != nil && *jobSpec.Completions < podsCount {
+		podsCount = *jobSpec.Completions
+	}
+	return podsCount
+}
+
+func generatePodsReadyCondition(job *batchv1.Job) metav1.Condition {
+	conditionStatus := metav1.ConditionFalse
+	message := "Not all pods are ready or succeeded"
+	if podsReady(job) {
```

Comment:
I was thinking about it. However, a suspended job does not need to block admission of new workloads as its pods are terminating, so will not deadlock. Also, a suspended job will eventually reach the PodsReady=False condition as its pods terminate (unless unsuspended quickly, but then we can save the two requests). 

Also, there might be a slight penalty to throuthput by an additional request to set `suspend=false` when a new workload is about to get admitted. WDYT?

Reply:
>  a suspended job does not need to block admission of new workloads 

That is true, but for that, we should be looking at `Workload.spec.admission != nil`.

> Also, a suspended job will eventually reach the PodsReady=False condition as its pods terminate (unless unsuspended quickly, but then we can save the two requests).

Since we are unsuspending after updating the jobs ready condition, I think what you have is fine.

Reply:
When the pods are finishing, then the Kueue Job reconciler will set `PodsReady=False` in response do decreased Ready count, but indeed it might be lagging a little bit. So what the scenario would be?
- workload A is admitted and the job A is unsuspended and gets all pods running (`PodsReady=True`)
- workload B preempts A, A gets suspended and removed from admitted workloads
- workload B completes very quickly, so that A is still in `PodsReady=True` ready to be re-admitted
- workload A gets readmitted with `PodsReady=True` even though it has a terminating pod now
- workload C gets admitted even though workload A still needs to restart some of its pods - both A and C are starting their pods at the same time.

However, I think the same scenario may theoretically still happen even if we set `PodsReady=False` when suspending. Still, it is unlikely, but it could happen that workload A gets re-admitted within the Job controller single syncJob pass thus, the pods will start to be terminated by invoking https://github.com/kubernetes/kubernetes/blob/0e19bbb91644885a6db38a77ea3d697730269802/pkg/controller/job/job_controller.go#L1416.

Both scenarios are very unlikely and with this feature we don't aim to give a full guarantee of no concurrent pods scheduling - this may still happen in more likely case of pod restart.

Reply:
Btw. I have just realized we don't add the Pod Disruption condition to pods deleted by the Job controller in reaction to suspending a workload. It seems like an omission / bug? I will create a bug if this is the case.

Reply:
Actually, after a second thought, I see that by setting `PodsReady=False` I would eliminate the risk of the scenario described above by being more conservative.

Reply:
It's not a "disruption". The user intention (setting `.spec.suspend=True`) is to terminate those pods.

Reply:
Pushed the suggested change PTAL.

Reply:
Maybe worth creating an issue for the discussion around job suspension and the disruption condition. Although, one reason why I thought is not a priority for Jobs is because these pods would get the finalizer removed before a condition is added and they wouldn't affect the failure policy.